finality-aleph/src/network/address_cache.rs

```diff
@@ -0,0 +1,184 @@
+    /// `session_info`. A session and validator_index identify the validator uniquely.
+    fn update(&self, validator_index: NodeIndex, session_info: ValidatorAddressingInfo);
+}
+
+enum ValidatorAddressCacheUpdaterImpl<K: KeyOwnerInfoProvider, A: AuthorityProvider> {
+    Noop,
+    BackendBased {
+        validator_address_cache: ValidatorAddressCache,
+        key_owner_info_provider: K,
+        authority_provider: A,
+        session_boundary_info: SessionBoundaryInfo,
+    },
+}
+
+/// Construct a struct that can be used to update `validator_address_cache`, if it is `Some`.
+/// If passed None, the returned struct will be a no-op.
+pub fn validator_address_cache_updater<K: KeyOwnerInfoProvider, A: AuthorityProvider>(
+    validator_address_cache: Option<ValidatorAddressCache>,
+    key_owner_info_provider: K,
+    authority_provider: A,
+    session_boundary_info: SessionBoundaryInfo,
+) -> impl ValidatorAddressCacheUpdater {
+    match validator_address_cache {
+        Some(validator_address_cache) => ValidatorAddressCacheUpdaterImpl::BackendBased {
+            validator_address_cache,
+            key_owner_info_provider,
+            authority_provider,
+            session_boundary_info,
+        },
+        None => ValidatorAddressCacheUpdaterImpl::Noop,
+    }
+}
+
+impl<K: KeyOwnerInfoProvider, A: AuthorityProvider> ValidatorAddressCacheUpdaterImpl<K, A> {
+    fn owner_at_session(
+        &self,
+        session: SessionId,
+        validator_index: NodeIndex,
+    ) -> Option<AccountId> {
+        match self {
+            ValidatorAddressCacheUpdaterImpl::Noop => None,
+            ValidatorAddressCacheUpdaterImpl::BackendBased {
+                key_owner_info_provider,
+                authority_provider,
+                session_boundary_info,
+                ..
+            } => {
+                let block_number = session_boundary_info
+                    .boundaries_for_session(session)
+                    .first_block();
+
+                let authority_data = authority_provider.authority_data(block_number)?;
+                let aleph_key = authority_data.authorities()[validator_index.0].clone();
+                key_owner_info_provider.aleph_key_owner(block_number, aleph_key)
+            }
+        }
+    }
+}
+
+impl<K: KeyOwnerInfoProvider, A: AuthorityProvider> ValidatorAddressCacheUpdater
+    for ValidatorAddressCacheUpdaterImpl<K, A>
+{
+    fn update(&self, validator_index: NodeIndex, info: ValidatorAddressingInfo) {
+        if let (
+            Some(validator),
+            ValidatorAddressCacheUpdaterImpl::BackendBased {
+                validator_address_cache,
+                ..
+            },
+        ) = (self.owner_at_session(info.session, validator_index), self)
+        {
+            validator_address_cache.insert(validator, info)
+        }
+    }
+}
+
+pub struct KeyOwnerInfoProviderImpl<C, B, BE>
```

Comment:
Since you are producing a struct here anyway, perhaps include session boundaries and the authority provider here as well (at least the latter is already implemented for this kind of client, innit?).

Alternatively, if you really want to keep these as separate pieces in the implementation above then maybe you can remove this struct and do the `impl` for just `Arc<C>`? Or is it not possible with the additional constraints?

Reply:
Also minor point – we try to keep the substrate-specific impls of these kinds of traits in separate files. I'm not sure if this is the best idea here (might make stuff more confusing for no good reason), but if you see an obvious way to make it happen that would be nice.